File: 3492-count-submatrices-with-equal-frequency-of-x-and-y/count-submatrices-with-equal-frequency-of-x-and-y.cpp
class Solution {
public:
    int numberOfSubmatrices(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        vector<vector<int>> x(m, vector<int>(n, 0));
        vector<vector<int>> y(m, vector<int>(n, 0));

        int ans = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {

                int currX = (grid[i][j] == 'X');
                int currY = (grid[i][j] == 'Y');

                x[i][j] = currX;
                y[i][j] = currY;

                if (i > 0) {
                    x[i][j] += x[i - 1][j];
                    y[i][j] += y[i - 1][j];
                }

                if (j > 0) {
                    x[i][j] += x[i][j - 1];
                    y[i][j] += y[i][j - 1];
                }

                if (i > 0 && j > 0) {
                    x[i][j] -= x[i - 1][j - 1];
                    y[i][j] -= y[i - 1][j - 1];
                }

                if (x[i][j] == y[i][j] && x[i][j] > 0) {
                    ans++;
                }
            }
        }

        return ans;
    }
};

File: 3492-count-submatrices-with-equal-frequency-of-x-and-y/count-submatrices-with-equal-frequency-of-x-and-y.java
class Solution {
    public int numberOfSubmatrices(char[][] grid) {
        int rows = grid.length, cols = grid[0].length;
        int[][] countX = new int[rows+1][cols+1];
        int[][] countY = new int[rows+1][cols+1];
        int ans = 0;
        for (int i = 1; i <= rows; i++) {
            for (int j = 1; j <= cols; j++) {
                int x = grid[i-1][j-1] == 'X' ? 1 : 0;
                int y = grid[i-1][j-1] == 'Y' ? 1 : 0;
                countX[i][j] = countX[i-1][j] + countX[i][j-1] - countX[i-1][j-1] + x;
                countY[i][j] = countY[i-1][j] + countY[i][j-1] - countY[i-1][j-1] + y;
                if (countX[i][j] == countY[i][j] && countX[i][j] > 0) ans++;
            }
        }
        return ans;
    }
}